signer/src/storage/postgres.rs

```diff
@@ -12,6 +52,111 @@ impl PgStore {
     pub async fn connect(url: &str) -> Result<Self, sqlx::Error> {
         Ok(Self(sqlx::PgPool::connect(url).await?))
     }
+
+    /// Write parts of the Stacks Nakamoto block headers to the database.
+    async fn write_stacks_block_header(&self, blocks: &[NakamotoBlock]) -> Result<(), Error> {
+        let block_summaries: Vec<StacksBlockSummary> = blocks
+            .iter()
+            .map(|block| StacksBlockSummary {
+                block_id: block.block_id().to_hex(),
+                chain_length: block.header.chain_length as i64,
+                parent_block_id: block.header.parent_block_id.to_hex(),
+            })
+            .collect();
+
+        if block_summaries.is_empty() {
+            return Ok(());
+        }
+
+        sqlx::query(
+            r#"
+            INSERT INTO sbtc_signer.stacks_blocks (block_hash, block_height, parent_hash, created_at)
+            SELECT
+                decode(block_id, 'hex')
+              , chain_length
+              , decode(parent_block_id, 'hex')
+              , CURRENT_TIMESTAMP
+            FROM JSONB_TO_RECORDSET($1::JSONB) AS x(
+                block_id        VARCHAR
+              , chain_length    BIGINT
+              , parent_block_id VARCHAR
+            )
+            ON CONFLICT DO NOTHING"#,
+        )
+        .bind(serde_json::to_string(&block_summaries).map_err(Error::JsonSerialize)?)
+        .execute(&self.0)
+        .await
+        .map_err(Error::SqlxQuery)?;
+
+        Ok(())
+    }
+
+    /// Write sBTC related transactions in the given blocks to the
+    /// database.
+    async fn write_stacks_sbtc_txs(&self, blocks: &[NakamotoBlock]) -> Result<(), Error> {
+        let transaction_kinds = contract_transaction_kinds();
+        let block_txs: Vec<StacksTx> = blocks
+            .iter()
+            .flat_map(|block| block.txs.iter().map(|tx| (tx, block.block_id())))
+            .filter_map(|(tx, block_id)| match &tx.payload {
+                TransactionPayload::ContractCall(x)
+                    if CONTRACT_NAMES.contains(&x.contract_name.as_str()) =>
+                {
+                    Some(StacksTx {
+                        tx_type: *transaction_kinds.get(&x.function_name.as_str())?,
+                        txid: tx.txid().to_hex(),
+                        block_id: block_id.to_hex(),
+                        tx: to_hex(&tx.serialize_to_vec()),
+                    })
+                }
+                _ => None,
+            })
+            .collect();
+
+        if block_txs.is_empty() {
+            return Ok(());
+        }
+
+        let block_txs_json = serde_json::to_string(&block_txs).map_err(Error::JsonSerialize)?;
+        sqlx::query(
```

Comment:
Nit: Again, I'd like to use the macro here if it's not too much work for you to change it.
```suggestion
        sqlx::query!(
```